src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -287,6 +308,66 @@ public TestResult findCorrespondingResult(String id) {
         return Collections.emptyList();
     }
 
+    /**
+     * Start for Test Result Trends.
+     * From here onwards start the code area responsible for generating various test result trends.
+     */
+
+    public hudson.tasks.junit.TestResult loadXml(){return null;}
+
+    /**
+     * A method for getting the list of packages for all levels of hierarchy.
+     * @return Array of packages of all hierarchies.
+     */
+    public String[] getProjectList(){
+
+        /*
+         * If project list is already ready no need to construct again.
+         */
+        if(projectList!=null) return projectList;
+        hudson.tasks.junit.TestResult r = loadXml();
+        Collection<SuiteResult> suiteList = r.getSuites();
+
+        /*
+         * A set for the package names.
+         */
+        Set<String> projectSet = new HashSet<String>();
+        for(SuiteResult sr: suiteList){
+            for(CaseResult cr: sr.getCases()){
+                String caseName = cr.getFullName();
+                String projectName = "";
+
+                /*
+                 * Splitting package by treating "." as separator.
+                 */
+                String[] packageTree = caseName.split("[.]");
+
+                /*
+                 * Iterating till the length-2 of packageTree array to exclude class name and test name.
+                 */
+                for(int i=0;i<packageTree.length-2;i++){
+                    if(!projectName.isEmpty()) projectName+='.';
+                    projectName+=packageTree[i];
+                    projectSet.add(projectName);
+                }
+            }
+        }
+
+        /*
+         * Converting the set of package names to the corresponding array
+         * and finally sorting it in ascending order.
+         */
+        int size = projectSet.size();
```

Comment:
you can use [Set.toArray()](https://docs.oracle.com/javase/6/docs/api/java/util/AbstractCollection.html#toArray(T[]))